nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java

```diff
@@ -0,0 +1,303 @@
+      if (missingFieldNames == null) {
+        message = "Method is annotated with @EnsuresNonNullIf but does not return boolean";
+      } else {
+        message =
+            String.format(
+                "Method is annotated with @EnsuresNonNullIf but does not ensure fields %s",
+                missingFieldNames);
+      }
+      state.reportMatch(
+          analysis
+              .getErrorBuilder()
+              .createErrorDescription(
+                  new ErrorMessage(ErrorMessage.MessageTypes.POSTCONDITION_NOT_SATISFIED, message),
+                  tree,
+                  analysis.buildDescription(tree),
+                  state,
+                  null));
+    }
+
+    // Clean up state again
+    semanticsHold = false;
+    missingFieldNames = null;
+    methodTreeUnderAnalysis = null;
+    methodSymbolUnderAnalysis = null;
+    returnTreesInMethodUnderAnalysis.clear();
+
+    return true;
+  }
+
+  private void buildUpReturnToEnclosingMethodMap(VisitorState methodState) {
+    returnTreesInMethodUnderAnalysis.clear();
+    new TreePathScanner<Void, Void>() {
+      @Override
+      public Void visitReturn(ReturnTree node, Void unused) {
+        VisitorState returnState = methodState.withPath(getCurrentPath());
+
+        TreePath enclosingMethod =
+            NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(returnState.getPath());
+        if (enclosingMethod == null) {
+          throw new RuntimeException("no enclosing method, lambda or initializer!");
+        }
+
+        // We only add returns that are directly in the method under analysis
+        if (!enclosingMethod.getLeaf().equals(methodTreeUnderAnalysis)) {
+          return super.visitReturn(node, null);
+        }
+
+        // We add the current return tree to the set as it's in the method
+        returnTreesInMethodUnderAnalysis.add(node);
+        return super.visitReturn(node, null);
+      }
+    }.scan(methodState.getPath(), null);
+  }
+
+  /*
+   * Sub-classes can only strengthen the post-condition.
+   * We check if the list in the child classes is at least the same as in the parent class.
+   */
+  @Override
+  protected void validateOverridingRules(
+      Set<String> overridingFieldNames,
+      NullAway analysis,
+      VisitorState state,
+      MethodTree tree,
+      Symbol.MethodSymbol overriddenMethod) {
+    FieldContractUtils.validateOverridingRules(
+        annotName, overridingFieldNames, analysis, state, tree, overriddenMethod);
+  }
+
+  @Override
+  public void onDataflowVisitReturn(
+      ReturnTree returnTree, NullnessStore thenStore, NullnessStore elseStore) {
+    // We only explore return statements that is inside
+    // the method under validation
+    if (!returnTreesInMethodUnderAnalysis.contains(returnTree)) {
+      return;
+    }
+
+    // We might have already found another return tree that results in
+    // what we need, so we don't keep going deep.
+    if (semanticsHold) {
+      return;
+    }
```

Comment:
I don't quite follow this.  Don't we need the fields to be non-null at _every_ return statement under the right condition?  I read this logic as checking if the fields are non-null (the semantics hold) for at least one statement.

Reply:
Old comment, not in the current implementation, closing it!